File: src/backend/vlm_feedback_loop/schemas/filesystem.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Pydantic schemas for filesystem browse and scan endpoints."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict

# ── Browse ──────────────────────────────────────────────────────────────────


class BrowseEntry(BaseModel):
    """Single entry (file or directory) in a browse response."""

    name: str
    type: Literal["directory", "file"]
    path: str
    size_bytes: int | None = None


class BrowseResponse(BaseModel):
    """Response body for ``GET /v1/filesystem/browse``."""

    model_config = ConfigDict(extra="forbid")

    path: str
    parent: str | None
    entries: list[BrowseEntry]
    bundled_sample_path: str | None = None


# ── Scan ────────────────────────────────────────────────────────────────────


class ScanRequest(BaseModel):
    """Request body for ``POST /v1/filesystem/scan``."""

    model_config = ConfigDict(extra="forbid")

    path: str
    recursive: bool = True
    project_id: str | None = None


class ScanImageEntry(BaseModel):
    """Single discovered image in a scan response."""

    storage_ref: str
    suggested_example_key: str
    size_bytes: int
    key_status: Literal[
        "available", "already_exists_same_path", "collision_different_path"
    ]
    existing_storage_ref: str | None = None


class ScanSkippedEntry(BaseModel):
    """A file skipped during scan with the reason."""

    path: str
    reason: str


class ScanResponse(BaseModel):
    """Response body for ``POST /v1/filesystem/scan``."""

    model_config = ConfigDict(extra="forbid")

    path: str
    images: list[ScanImageEntry]
    skipped: list[ScanSkippedEntry]
    total_images: int
    total_skipped: int
    total_collisions: int

File: src/ui/src/types/filesystem.ts
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * TypeScript mirrors of backend filesystem and ingestion schemas.
 */

// ── Browse ──────────────────────────────────────────────────────────────────

export interface BrowseEntry {
  name: string;
  type: "directory" | "file";
  path: string;
  size_bytes: number | null;
}

export interface BrowseResponse {
  path: string;
  parent: string | null;
  entries: BrowseEntry[];
  bundled_sample_path: string | null;
}

// ── Scan ────────────────────────────────────────────────────────────────────

interface ScanImageEntry {
  storage_ref: string;
  suggested_example_key: string;
  size_bytes: number;
  key_status: "available" | "already_exists_same_path" | "collision_different_path";
  existing_storage_ref: string | null;
}

interface ScanSkippedEntry {
  path: string;
  reason: string;
}

export interface ScanResponse {
  path: string;
  images: ScanImageEntry[];
  skipped: ScanSkippedEntry[];
  total_images: number;
  total_skipped: number;
  total_collisions: number;
}

// ── Ingest ──────────────────────────────────────────────────────────────────

interface IngestItem {
  example_key: string;
  storage_ref: string;
  source_metadata?: Record<string, unknown>;
}

export interface IngestRequest {
  examples: IngestItem[];
}

interface IngestResultItem {
  example_key: string;
  status: "created" | "exists" | "error";
  error: string | null;
  error_code: string | null;
  warnings: string[];
  example: Record<string, unknown> | null;
}

export interface IngestResponse {
  results: IngestResultItem[];
}
